File: rl/markov_action_generator.py
import numpy as np


class MarkovActionGenerator:
    def __init__(self):
        self.chance_of_random_action = .3
        self.none = np.array([1., 0., 0., 0., 0.])
        self.left = np.array([0., 1., 0., 0., 0.])
        self.up = np.array([0., 0., 1., 0., 0.])
        self.right = np.array([0., 0., 0., 1., 0.])
        self.down = np.array([0., 0., 0., 0., 1.])
        self.action = self.random_action()

    def random_action(self):
        r = np.random.random()
        if r < .2:
            return self.none
        elif r < .4:
            return self.left
        elif r < .6:
            return self.up
        elif r < .8:
            return self.right
        else:
            return self.down

    def next_action(self):
        if np.random.random() < self.chance_of_random_action:
            self.action = self.random_action()
        return self.action

File: web/markov_action_generator.js
class MarkovActionGenerator {
  constructor() {
    this.chance_of_random_action = 0.3;
    this.none = [1, 0, 0, 0, 0];
    this.left = [0, 1, 0, 0, 0];
    this.up = [0, 1, 1, 0, 0];
    this.right = [0, 0, 0, 1, 0];
    this.down = [0, 0, 0, 0, 1];
    this.action = this.random_action();
  }

  random_action() {
    const r = Math.random()
    if (r < 0.2) {
      return this.none;
    } else if (r < 0.4) {
      return this.left;
    } else if (r < 0.6) {
      return this.up
    } else if (r < 0.8) {
      return this.right;
    } else {
      return this.down;
    }
  }

  get_action() {
    if (Math.random() > this.chance_of_random_action) {
      this.action = this.random_action();
    }
    return this.action;
  }
}
